### ISSUE TITLE
TSDS data streams can route writes to downsampled indices after rollover

### ISSUE THREAD

**Issue by andreidan** [elastic/elasticsearch#99696]
### Elasticsearch Version

8.10, 8.11

### Java Version

_bundled_

### OS Version

Darwin

### Problem Description

A tsds backing index has configured a `START_TIME` and `END_TIME` denoting the time bounds for the data they will host. The `END_TIME` in particular is configured based on the `index.look_ahead_time` setting.

All writes against a TSDS will be routed based on the document `@timestamp` to the correct backing index according to each index's START/END time configuration.

We can easily simulate a situation where we rollover the data stream and the `now` timestamp will now not be routed to the write index anymore but to the second generation index (because the second generation index's `END_TIME` configuration has not lapsed yet). This would normally not be a problem but if the index is read-only the write will fail. 

The index could be read-only because it was downsampled. For downsampling in particular we should delay downsampling the backing index until the configured `END_TIME` for the backing index has lapsed. Note that a similar situation could be encountered if, say, a `searchable_snapshot` action is used instead of `downsampling` - however we should probably treat that separately.

This problem is present both in ILM and data stream lifecycle.

In ILM, currently, the only workaround is to increase the `min_age` of the phase where downsampling is configured, sufficiently such that the `look_ahead_time` for the backing indices will have lapsed by the time the index transitions to the phase where downsampling is configured. 

### Steps to Reproduce

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "5s"
  }
}


PUT _component_template/test-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "format": "epoch_millis",
          "type": "date"
        },
        "metricKey": {
          "time_series_dimension": true,
          "type": "keyword"
        },
        "value": {
          "time_series_metric": "gauge",
          "type": "float"
        }
      }
    }
  }
}

PUT _component_template/test-settings
{
  "template": {
    "settings": {
      "index": {
        "lifecycle": {
          "name": "test-lifecycle"
        },
        "look_ahead_time": "10m"
      }
    }
  }
}

PUT _index_template/test1-template
{
  "priority": 500,
  "template": {
    "settings": {
      "index": {
        "mode": "time_series"
      }
    }
  },
  "index_patterns": [
    "test1*"
  ],
  "data_stream": {
    "hidden": false,
    "allow_custom_routing": false
  },
  "composed_of": [
    "test-settings",
    "test-mappings"
  ]
}

PUT _ilm/policy/test-lifecycle
{
  "policy": {
    "phases": {
      "hot": {
        "min_age": "0ms",
        "actions": {
          "downsample": {
            "fixed_interval": "5m"
          },
          "rollover": {
            "max_primary_shard_size": "25gb",
            "max_age": "1h",
            "max_docs": 2
          },
          "set_priority": {
            "priority": 100
          }
        }
      },
      "delete": {
        "min_age": "7d",
        "actions": {
          "delete": {
            "delete_searchable_snapshot": true
          }
        }
      }
    }
  }
}


===============
Get `now` date
❯ date +"%s%3N"
1695131067275
================

POST _bulk
{"create": {"_index": "test1ds2"}}
{"metricKey": "thekey1", "@timestamp": "1695131067275", "value": 1, "instance": "server:9100", "job": "node", "env": "hml", "__name__": "up"}

== rollover the data stream
POST test1ds2/_rollover

==Get `now` date
❯ date +"%s%3N"
1695131118582

POST _bulk
{"create": {"_index": "test1ds2"}}
{"metricKey": "thekey2", "@timestamp": "1695131118582", "value": 1, "instance": "server:9100", "job": "node", "env": "hml", "__name__": "up"}
```

The result is
```
{
  "errors": true,
  "took": 0,
  "items": [
    {
      "create": {
        "_index": "test1ds2",
        "_id": null,
        "status": 403,
        "error": {
          "type": "cluster_block_exception",
          "reason": "index [downsample-5m-.ds-test1ds2-2023.09.19-000001] blocked by: [FORBIDDEN/8/index write (api)];"
        }
      }
    }
  ]
}
```

**Comment by andreidan** [comment]
The ILM actions that should probably wait for the `end_time` to laps are 
`downsample`, `readonly`, `searchable_snapshot` and `shrink`

**Comment by CyberTaoFlow** [comment]
got hit by this just today.  ILM that instigated it is below.  Also we are using logstash between agents and ingest.   The 403 handling (retry forever) brought our pipeline to a halt basically.

{
  "policy": {
    "phases": {
      "hot": {
        "min_age": "0ms",
        "actions": {
          "rollover": {
            "max_age": "7d",
            "max_primary_shard_size": "50gb"
          },
          "set_priority": {
            "priority": 100
          },
          "downsample": {
            "fixed_interval": "1h"
          }
        }
      },
      "warm": {
        "min_age": "1d",
        "actions": {
          "downsample": {
            "fixed_interval": "1d"
          },
          "set_priority": {
            "priority": 50
          },
          "allocate": {
            "number_of_replicas": 1
          }
        }
      },
      "cold": {
        "min_age": "30d",
        "actions": {
          "set_priority": {
            "priority": 0
          },
          "migrate": {
            "enabled": false
          },
          "allocate": {
            "number_of_replicas": 0
          }
        }
      },
      "delete": {
        "min_age": "60d",
        "actions": {
          "delete": {
            "delete_searchable_snapshot": true
          }
        }
      }
    }
  }
}

**Comment by andreidan** [comment]
Removing `team-discuss` as we discussed this and decided we'll add an extra step in these following actions: 
```
downsample, readonly, searchable_snapshot, shrink
```

This step will wait for the `index.time_series.end_time` to lapse before proceeding with executing the action. We'll document this behaviour both in the ILM docs and the downsampling docs (cross referencing to ILM). 

Additionally, the explain API will provide a useful indication when an index waits in for the time to lapse as part of the existing `step_info` field.

**Comment by mbudge** [comment]
As we have 1 pipeline for elasticagent in Logstash this knocked all our production integrations/endpoint log collection offline, which caused a 7 day outage while I was on holiday. Logstash got stuck in a retry loop as it couldn't index to the read only downsample indexes. Please make sure TSDB/downsampling works with Logstash, or if there is anything we can catch us out it is clearly documented.

**Comment by StephanErb** [comment]
> Removing `team-discuss` as we discussed this and decided we'll add an extra step in these following actions: 
> ```
> downsample, readonly, searchable_snapshot, shrink
> ```
@andreidan  please also add forcemerge to the list. We are doing a forcemerge to 1 on the hot tier, but end up with more segments as indexing continues after the forcemerge completes.

**Comment by andreidan** [comment]
@StephanErb ++ 
Thanks, that's a great suggestion.

**Comment by andreidan** [comment]
Reopening this as only ILM was fixed. Data stream lifecycle is still left to be fixed.

**Comment by andreidan** [comment]
@mbudge sorry to hear about the incident you experienced. This has been fixed within ILM, could you please post your ILM policy here to double check?

**Comment by mbudge** [comment]
@andreidan it's okay checked the policy with support. 

We'll test downsampling again after a few more releases.

**Comment by mbudge** [comment]
We had another epic outage caused by downsampling, even though it's disabled in ILM and we had deleted the downsampling indexes. 

Please release 8.10.3 asap.

**Comment by taremaa** [comment]
I found this issue after encountering the same bug last week.

If I understand correctly, the linked PR excludes the backing indices that have `end_time` greater than or equal to `now()`. After that, the indices may be downsampled (shrunk, made read only, etc) immediately. Doesn't this still leave a smaller issue of data being ingested with a delay and being rejected because the backing index with the matching range has already been downsampled?

For example, if the timestamp is `near_end_time` and the document is being ingested at `near_end_time + metric_pipeline_latency` > `end_time`, the backing index matching the timestamp might already be downsampled.

The documentation on this topic has left me an impression that the data should be accepted as long as the timestamp is greater than `now - index.look_ahead_time`: [Accepted time range for adding data](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html#tsds-accepted-time-range)

It may make sense to tackle this as a separate bug, though.

**Comment by andreidan** [comment]
Note that this was fixed for ILM in `8.10.3` and for both ILM and DSL in `8.11`

**Comment by mbudge** [comment]
We had another outage in December. Something caused metrics data to arrive late.

elastic returned 409 Forbidden as the index had rolled over to frozen by the time the metrics data arrived.

Logstash got stuck retrying the metrics data indefinitely. 

Recommend a filter in logstash to drop 409 events (not dead letter queue), or add a processor to Fleet managed metrics ingest pipelines which drops metrics data which is over 24 hours late.

https://github.com/elastic/logstash/issues/15832
